### PR TITLE
Standard Snippets Improvements

### PR DESCRIPTION
**Pull request by dharmrajsinh-jhala**
Description of the issue/feature this PR addresses:
Task : https://www.odoo.com/web#id=30739&view_type=form&model=project.task&action=327&menu_id=4720
Pad : https://pad.odoo.com/p/r.d086d43cffbb8df0210231d2ce192036

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website/views/snippets.xml [hunk 1]
```diff
@@ -365,6 +365,74 @@
...
+<template id="s_company_team" name="Company Team">
+    <section class="s_company_team">
+        <div class="container">
+            <div class="row mb32">
+                <div class="col-md-12 col-xs-12">
+                    <h1 class="text-center">
+                        Meet the Executive Team
+                    </h1>
+                </div>
+            </div>
```
Comment (qsm-odoo) [inline addons/website/views/snippets.xml]: Pretty sure the `col-xs-12` (and all of them below) is useless. Also are these resizable titles back? :) I had to remove them a while ago... :)

File: addons/website/views/snippets.xml [hunk 2]
```diff
@@ -528,7 +602,7 @@
 </template>
 
 <template id="s_parallax" name="Parallax">
-    <section class="parallax s_parallax oe_custom_bg oe_img_bg"
+    <section class="parallax s_parallax oe_img_bg"
```
Comment (qsm-odoo) [inline addons/website/views/snippets.xml]: Sure about this?

File: addons/website/views/snippets.xml [hunk 3]
```diff
@@ -365,6 +365,74 @@
...
+            <div class="row mb32">
+                <div class="col-md-12 col-xs-12">
+                    <h1 class="text-center">
+                        Meet the Executive Team
+                    </h1>
+                </div>
+            </div>
+            <div class="row">
+                <div class="col-md-6 col-xs-12 mb32">
+                    <div class="col-md-4">
```
Comment (qsm-odoo) [inline addons/website/views/snippets.xml]: There is a row missing here

File: addons/web_editor/static/src/js/snippets.editor.js [hunk 4]
```diff
@@ -961,6 +961,9 @@ data.Editor = Class.extend({
             this.$overlay.find(".oe_snippet_move, .oe_snippet_clone").addClass('hidden');
         }
 
+        if (!$styles.find('li.dropdown-submenu').length && !$styles.find('li.o_parent_editor_header').length) {
+            $styles.hide();
+        }
```
Comment (dja-odoo) [inline addons/web_editor/static/src/js/snippets.editor.js]: Done

File: addons/web_editor/static/src/js/snippets.editor.js [hunk 5]
```diff
@@ -994,7 +997,13 @@ data.Editor = Class.extend({
...
         this.buildingBlock.call_for_all_snippets($clone, function (editor, $snippet) {
             for (var i in editor.styles) {
-                editor.styles[i].on_clone($snippet);
+                var auxiliary_options = {'is_current': false};
+                // Below condition is added to remove offset of target element only
+                // and not its children to avoid design alteration of a container / block.
+                if ($snippet.is($clone)){
+                    auxiliary_options = {'is_current': true};
+                }
+                editor.styles[i].on_clone($snippet, auxiliary_options);
```
Comment (dja-odoo) [inline addons/web_editor/static/src/js/snippets.editor.js]: Done

File: addons/web_editor/static/src/js/snippets.options.js [hunk 6]
```diff
@@ -576,15 +576,18 @@ odoo.define('web_editor.snippets.options', function (require) {
...
             return res;
         },
 
+        bind_bg_events: function () {
+            this.$target.off(".background-option")
+            .on("background-color-event.background-option", (function (e, type) {
+                e.stopPropagation();
+                if (e.currentTarget !== e.target) return;
+                this.$el.find("li:first > a").trigger(type);
+            }).bind(this));
```
Comment (dja-odoo) [inline addons/web_editor/static/src/js/snippets.options.js]: Done

File: addons/website/static/src/js/website.snippets.editor.js [hunk 7]
```diff
@@ -218,6 +218,7 @@ options.registry.carousel = options.registry.slider.extend({
 
         // set background and prepare to clean for save
         this.$target.on('slid.bs.carousel', function () {
+            if (!self.editor.styles.background) return;
```
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.editor.js]: Done

File: addons/website/static/src/js/website.snippets.editor.js [hunk 8]
```diff
@@ -345,7 +352,6 @@ options.registry.parallax = options.Class.extend({
     _refresh: function () {
         _.defer((function () {
             this.$target.data("snippet-view")._rebuild();
-            this._update_target_to_bg();
         }).bind(this));
```
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.editor.js]: Yes. This fixes the issue of Parallax's background being changed when just browsing ( without clicking ) available Background Images from drop-down.

File: addons/website/models/ir_ui_view.py [hunk 9]
```diff
@@ -120,11 +120,21 @@ def _prepare_qcontext(self):
...
         editable = not translatable and editable
 
+        def unslug(s):
+            parts = s.split('/')
+            if parts:
+                unslug_val = website.unslug(parts[-1])
+                if unslug_val[1]:
+                    parts[-1] = str(unslug_val[1])
+                    return '/'.join(parts)
+            return s
```
Comment (dja-odoo) [inline addons/website/models/ir_ui_view.py]: Done

File: addons/website/static/src/js/website.snippets.animation.js [hunk 10]
```diff
@@ -328,7 +328,7 @@ animation.registry.parallax = animation.Class.extend({
         if (!this.$bg || !this.$bg.length) {
             this.$bg = this.$("> .s_parallax_bg");
             if (!this.$bg.length) {
-                this.$bg = $("<span/>", {"class": "s_parallax_bg"}).prependTo(this.$target);
+                this.$bg = $("<span/>", {"class": this.$target.hasClass('oe_custom_bg') ? "s_parallax_bg oe_custom_bg" : "s_parallax_bg"}).prependTo(this.$target);
```
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.animation.js]: To prevent losing custom background on parallax slider ( issue : https://youtu.be/JNfBWiR6ffM?t=21s )

File: addons/website/static/src/less/website.snippets.less [hunk 11]
```diff
@@ -11,6 +11,15 @@
...
 }
 
+// Inner content Snipptes
+#wrapwrap
+    .row
+        [class^="col-md-"]
+            .s_auto_height
+            {
+                height: auto;
+            }
```
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: Nope in theme_common, height is 100% for many snippets. So this class is added to the inner content snippets (to prevent issues like : https://drive.google.com/file/d/0B1uIL9E_zXrrNmZSZGZ5dDZvdFU/view )

File: addons/website/static/src/less/website.snippets.less [hunk 12]
```diff
@@ -53,9 +62,33 @@ table.table_desc tr td {
...
+    }
+    .text-muted {
+        color: @text-muted;
+    }
+}
+
+.s_faq_collapse .container .panel {
+    background-color: inherit;
+    border : 1px solid rgba(0, 0, 0, 0.15);
+    & .panel-body {
```
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: Done

File: addons/website/views/snippets.xml (hunk 1 above)
Comment (dja-odoo) [inline addons/website/views/snippets.xml]: Done. Removed `col-xs-12` and resizable Title :)

File: addons/website/views/snippets.xml (hunk 2 above)
Comment (dja-odoo) [inline addons/website/views/snippets.xml]: Yes because oe_custom_bg is for customly selected images (selected from the image dialog box) and by default, this snippet is having background selected from the available images' dropdown (cubes) and not custom one.

File: addons/website/views/snippets.xml (hunk 3 above)
Comment (dja-odoo) [inline addons/website/views/snippets.xml]: Actually it's a tweak to select image and text block together ( like this : https://goo.gl/gzrVhO ) so that it's convenient for end user to delete / duplicate the whole block while keeping it responsive.

File: addons/website/static/src/less/website.snippets.less (hunk 11 above)
Comment (qsm-odoo) [inline addons/website/static/src/less/website.snippets.less]: I don't think this is useful (the least classes we have the better). The issue you showed is a theme issue. There is a rule which says:
```less
#wrapwrap .s_four_columns_fw div {
    height: 100%;
}
```
which is wrong.
It should be more precise so that the 100% rule applies only on the own snippet div elements (`> .container-fluid, .container-fluid > .row, .container-fluid > .row > div`).

File: addons/website/views/snippets.xml (hunk 3 above)
Comment (qsm-odoo) [inline addons/website/views/snippets.xml]: Mmh ok, but I think that XML structure is more important. What you could do is put the row with a `s_no_resize_cols` class and change the `margin-x` option so that the `data-selector` is `.row:not(.s_no_resize_cols) > [class*='col-md-']`.

File: addons/web_editor/static/src/js/snippets.editor.js [hunk 13]
```diff
@@ -994,7 +994,9 @@ data.Editor = Class.extend({
         this.$target.after($clone);
         this.buildingBlock.call_for_all_snippets($clone, function (editor, $snippet) {
             for (var i in editor.styles) {
-                editor.styles[i].on_clone($snippet);
+                editor.styles[i].on_clone($snippet, {
+                    isCurrent: ($snippet.is($clone))
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/snippets.editor.js]: We are writing JS guidelines here: https://github.com/odoo/odoo/wiki/Javascript-coding-guidelines
Here there is a missing trailing comma.
So:
```javascript
{
    isCurrent: $snippet.is($clone),
}
```

File: addons/website/static/src/js/website.snippets.animation.js (hunk 10 above)
Comment (qsm-odoo) [inline addons/website/static/src/js/website.snippets.animation.js]: I think this would be more clear like this then:
```javascript
this.$bg = $('<span/>', {
    class: 's_parallax_bg' + (this.$target.hasClass('oe_custom_bg') ? '' : ' oe_custom_bg'),
}).prependTo(this.$target);
```

File: addons/website/static/src/js/website.snippets.editor.js [hunk 14]
```diff
@@ -233,6 +233,9 @@ options.registry.carousel = options.registry.slider.extend({
                 s_option.$target = self.$target.find(".item.active");
                 s_option.set_active();
                 s_option.$target.trigger("snippet-option-change", [s_option]);
+                if (opt_name == 'background') {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/website.snippets.editor.js]: Just convention: use `===`

File: addons/website/static/src/js/website.snippets.gallery.js [hunk 15]
```diff
@@ -186,6 +185,7 @@ options.registry.gallery = options.Class.extend({
         this.$target.css("height", "");
     },
     slideshow :function (type) {
+        var imgStyle = this.$el.find('li.active[data-styling]').data('styling') || ''
```
Comment (qsm-odoo) [inline addons/website/static/src/js/website.snippets.gallery.js]: Missing `; `

File: addons/website/static/src/js/website.snippets.animation.js (hunk 10 above)
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.animation.js]: Done

File: addons/website/static/src/less/website.snippets.less (hunk 11 above)
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: Done. Fixed in theme snippets

File: addons/website/views/snippets.xml (hunk 3 above)
Comment (dja-odoo) [inline addons/website/views/snippets.xml]: Done

File: addons/web_editor/static/src/js/snippets.editor.js (hunk 13 above)
Comment (dja-odoo) [inline addons/web_editor/static/src/js/snippets.editor.js]: Done

File: addons/website/static/src/js/website.snippets.editor.js (hunk 14 above)
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.editor.js]: Done

File: addons/website/static/src/js/website.snippets.editor.js [hunk 16]
```diff
@@ -265,9 +268,13 @@ options.registry["margin-x"] = options.registry.marginAndResize.extend({
 
         return this.grid;
     },
-    on_clone: function ($clone) {
-        var _class = $clone.attr("class").replace(/\s*(col-lg-offset-|col-md-offset-)([0-9-]+)/g, '');
-        $clone.attr("class", _class);
+    on_clone: function ($clone, auxiliaryOptions) {
+        // Below condition is added to remove offset of target element only
+        // and not its children to avoid design alteration of a container / block.
+        if (auxiliaryOptions.isCurrent) {
```
Comment (qsm-odoo) [inline addons/website/static/src/js/website.snippets.editor.js]: In fact, just name the variable `options`... :)

File: addons/website/static/src/less/website.snippets.less [hunk 17]
```diff
@@ -53,9 +53,32 @@ table.table_desc tr td {
 }
 
 // Other snippet styles
+.s_feature_grid p{
+    overflow: hidden;
+}
```
Comment (qsm-odoo) [inline addons/website/static/src/less/website.snippets.less]: `overflow: hidden;` is nearly always bad, why is this necessary here ? :)

File: addons/website/static/src/less/website.snippets.less [hunk 18]
```diff
@@ -53,9 +53,32 @@ table.table_desc tr td {
...
+
 .s_comparisons .panel .list-group-item:not(.active) {
     color: #666;
 }
+.s_comparisons {
+    .panel small {
+        color:inherit;
```
Comment (qsm-odoo) [inline addons/website/static/src/less/website.snippets.less]: missing space, but why is that rule necessary ?

File: addons/website/static/src/less/website.snippets.less [hunk 19]
```diff
@@ -218,25 +241,25 @@ div.carousel .container {
...
             margin: 10px 0;
-            * {
+            *:not(img) {
                 padding: 0 10px;
             }
         }
         &.o_spc-big div.row {
             margin: 15px 0;
-            * {
+            *:not(img) {
```
Comment (qsm-odoo) [inline addons/website/static/src/less/website.snippets.less]: I think that `> div` would be better than `*:not(img)` for all of these

File: addons/website/views/snippets.xml [hunk 20]
```diff
@@ -365,6 +365,80 @@
...
 </template>
 
+<template id="s_company_team" name="Company Team">
+    <section class="s_company_team">
+        <div class="container">
+            <div class="row mb32">
+                <h1 class="text-center">
+                    Meet the Executive Team
+                </h1>
+            </div>
```
Comment (qsm-odoo) [inline addons/website/views/snippets.xml]: There can't be content in a `.row` which is not a `.col-*`. Here I think you should simply put the `h1` element as direct child of the `.container` (and put the `mb32` class on it if necessary).

File: addons/website/static/src/less/website.snippets.less (hunk 17 above)
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: To prevent text coming below the icon / image like this : 
feature-grid-overflow

File: addons/website/static/src/js/website.snippets.editor.js (hunk 16 above)
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.editor.js]: Done :)

File: addons/website/static/src/less/website.snippets.less (hunk 18 above)
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: Because, by default small tag is [grey](https://github.com/odoo/odoo/blob/master/addons/web/static/lib/bootstrap/css/bootstrap.css#L1218) colored in bootstrap, which causes this issue : 
compare-small-color

File: addons/website/static/src/less/website.snippets.less (hunk 19 above)
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: Done

File: addons/website/views/snippets.xml (hunk 20 above)
Comment (dja-odoo) [inline addons/website/views/snippets.xml]: Right :+1:  Done :)

File: addons/website/static/src/js/website.snippets.gallery.js (hunk 15 above)
Comment (dja-odoo) [inline addons/website/static/src/js/website.snippets.gallery.js]: Done

File: addons/website/static/src/less/website.snippets.less [hunk 21]
```diff
@@ -53,9 +53,32 @@ table.table_desc tr td {
 }
 
 // Other snippet styles
+.s_feature_grid p {
+    overflow: hidden;
```
Comment (qsm-odoo) [inline addons/website/static/src/less/website.snippets.less]: Mmh ok. I think this should work with `overflow: auto;` which is less "dangerous". Also, add a comment to explain the problem (or simply add "force the creation of a block formatting context").

**Comment by qsm-odoo** [comment]
I tested, that's great. Still another small problem though, tested on Firefox (the text is white on white in the default snippet) : 
s_faq_collapse

File: addons/website/static/src/less/website.snippets.less (hunk 21 above)
Comment (dja-odoo) [inline addons/website/static/src/less/website.snippets.less]: Done

**Comment by dja-odoo** [comment]
IDK if issue with the [white fonts](https://cloud.githubusercontent.com/assets/10338094/26625513/9d6be7b4-45f4-11e7-8e2c-b7eb21676ca2.png) is of runbot or not. Working well after rebase.

**Comment by qsm-odoo** [comment]
Merged.
